### ISSUE TITLE
Clickjacking vulnerability

### ISSUE THREAD

**Issue by marektoth** [keepassxreboot/keepassxc-browser#1367]
KeePassXC-Browser is vulnerable to clickjacking because it is possible to hide the icon or parts of the extension. 

A user visits the website for which he has a saved password. If the website is vulnerable to XSS (it doesn't have to be a login page) and the attacker has already exploited it, as soon as the user clicks anywhere, his saved password will be stolen.
## Current Behavior
In the KeepassXC-Browser the password is filled in by default when you click on the icon.
If a new form is created with `opacity:0.2`, the icon is still same visible.
![](https://marektoth.com/images/keepassxc-opacity.png)
But if the form is in an iframe and the iframe has `opacity:0.2`, the icon is transparent in the frame => possible to make the icon invisible.
![](https://marektoth.com/images/keepassxc-click2.gif)
The attack consists of an attacker exploiting an XSS vulnerability, creating a new login form, and then loading the same site in an iframe. The content in the iframe will be shifted to the newly created form so that the user always clicks on the icon. 

Examples of invisible iframe attacks:
1. An attacker put an invisible iframe where the user probably clicks
2. (Demo is available below) Invisible iframe under the mouse cursor - An attacker creates an invisible iframe with the size of KeepassXC-Browser icon. Content in the iframe would be exactly positioned on the icon and the iframe will tracks the position of the cursor. In this case, wherever the user clicks, he always clicks on the icon -> login data will be filled.

The general problem, in this case, is that the extension is not always "on-top" in the iframe. In addition to the mentioned making the icon invisible, it is possible to cover parts of the extension (the icon) or affect the visible part (change visible size). Below you can see the reduced width of the iframe that affects the visible part of the icon.
![](https://marektoth.com/images/keepassxc-iframesize.png)
Another way of abuse would be to have a 1x1 px iframe and the content would be targeted exactly to the icon. Because of its size, this pixel iframe would not be visible even if the opacity:1 is set. The iframe would always be positioned under the cursor and the technique would be the same - wherever the user clicks, he always clicks on the icon.

If more logins are stored on the domain, the user is shown a selection menu. In the iframe, this selection menu is also hidden for the user. Abuse of KeepassXC-Browser is possible even if multiple logins are stored.
![](https://marektoth.com/images/KeepassXC-logins.png)
## Expected Behavior
Extension elements should be "on top" and can not be changed, for example transparency or size.

## Attack Scenario
1) An attacker creates via XSS exactly same login form that is used on the domain
2) Then he creates an iframe that will load the same website
3) The content in the iframe will be positioned where the icon is located
4) The iframe will have opacity:0 - icon is invisible but is clickable
5) The new login form has event onchange() and if user click on invisible icon -> form is changed and attacker receive credentials

## Steps to Reproduce
Demo (invisible icon): 
1) Save your password on this site: [URL](https://websecurity.dev/password-managers/login/) 
2) Go to URL and click somewhere: [URL](https://websecurity.dev/password-managers/clickjacking/)
3) Your credentials will be in alert

[Video](https://youtu.be/iNKa8GCIk7o)
## Debug info
KeePassXC - 2.6.6
KeePassXC-Browser - 1.7.8.1
Operating system: Windows/Mac
Browser: Firefox/Chrome/Chromium

**Comment by droidmonkey** [comment]
If the iframe loads a different url than the one the credentials are stored in then nothing will be filled in.

**Comment by marektoth** [comment]
For example, if someone finds XSS stored on a website he can use this method. With the XSS vulnerability, he only has client-side access, but using this vulnerability he can easily steal users' passwords.

**Comment by varjolintu** [comment]
<img width="838" alt="Screenshot 2021-07-12 at 17 30 41" src="https://user-images.githubusercontent.com/24570482/125305324-f00c2900-e336-11eb-9a79-04c0c58b711e.png">

The test page itself is not very user friendly. This dialog cannot be dismissed. None of the buttons work.

**Comment by varjolintu** [comment]
I made a fix where form or input fields inside forms with minimal opacity are ignored. At least that doesn't allow the credential retrieval.

Still, the hijack would be only possible if the hijacker is using an iframe from the same site and domain. The attacker would have a lot more than a few credentials.

**Comment by marektoth** [comment]
Yes, I know, it's just a demo. But you should still see your saved credentials if you click anywhere on the website. Of course, you must first save the password on the test login form.

If only the opacity value is checked, it can still be abused in the way that the iframe is resized to 1x1 px (with opacity:1) and the content in the iframe is again positioned on the icon.

I agree, an attacker can definitely would have a lot more if he finds an XSS vulnerability, but he will never (easily) get the victim's password in plain text.

**Comment by varjolintu** [comment]
The attack surface can be reduced a lot if user doesn't allow access to all credentials automatically, and keeps the manual confirmation enabled, even if it's not as convenient.

It's also a good thing to remember that in general browsers don't allow injecting extension content scripts to iframes that are from another domain. I'd really like to see a "real" demo of this kind of hijacking.

**Comment by marektoth** [comment]
Currently, KeePassXC-Browser fills in passwords on click, which is quite enough interaction from the user. Problematically, this can be exploited, because the extension is not "on top" in the iframe. So currently, clickjacking can be abused with KeePassXC-Browser in the default configuration.

This vulnerability is part of the analysis on the password manager that I plan to publish today https://marektoth.com/blog/password-managers-autofill/. It may make more sense in the whole context and not just part. It was focused on autofill, but currently, with KeePassXC-Browser it is possible to get a password in one click (same as chromium-based browser).

_"..browsers don't allow injecting extension content scripts to iframes that are from another domain."_ I know and that's not the case. Let's say an attacker finds Stored XSS on amazon.com. The vulnerability will be in the product review - so the attacker can insert a malicous script for each product. The XSS is a client-side vulnerability, so the attacker can manipulate with the victim's account, but he can't find out the password - he can only change it (if enabled). By exploiting an autofill in password management or this vulnerability, an attacker can obtain the passwords of users in plaintext for a domain. When news comes out that 100 K users logins in plaintext leaked from Amazon. It doesn't necessarily mean that an attacker got into the database at all, but that he just exploited a misconfiguration in password managers (autofill) or this type of vulnerability.

To summarize, the attacker's goal is not to display an iframe on another domain, but on the domain where he wants to steal user credentials in plaintext. And the victim doesn't have to log in to the website again or fill in a fake dialog, just click somewhere.

**Comment by varjolintu** [comment]
Before the interaction, KeePassXC still shows an access confirmation dialog by default if input fields are detected and credentials needs to be retrieved. If there's a page you cannot even see input fields and you got the dialog (or you are already logged in, who's asking them again?), it's obvious to deny the access.

**Comment by marektoth** [comment]
Confirmation dialog is not shown by default. The video https://youtu.be/iNKa8GCIk7o shows the default configuration with the test site.

**Comment by varjolintu** [comment]
Ah yes. If you create an entry using the extension, the host is automatically allowed. Maybe we should make that optional. If you just create an entry using KeePassXC, the confirmation dialog will be shown.

**Comment by marektoth** [comment]
Oh, you're right. I only tested KeePassXC-Browser. If someone is using the browser integration, then he is more likely to use adding credentials via the extension than KeePassXC directly.

**Comment by GodSaveTheDoge** [comment]
> Ah yes. If you create an entry using the extension, the host is automatically allowed. Maybe we should make that optional. If you just create an entry using KeePassXC, the confirmation dialog will be shown.

Any updates on this? The behavior hasn't changed by the look of things.

**Comment by varjolintu** [comment]
> > Ah yes. If you create an entry using the extension, the host is automatically allowed. Maybe we should make that optional. If you just create an entry using KeePassXC, the confirmation dialog will be shown.
> 
> Any updates on this? The behavior hasn't changed by the look of things.

Not yet.

**Comment by schwedenmut** [comment]
Even though the new version 1.9.9.2 and the changes from #2641 are now in place, it seems that this still is vulnerable to the communicated flaws at defcon 33 (cp. https://marektoth.com/blog/dom-based-extension-clickjacking/#keepassxc-browser with the poc https://websecurity.dev/video/keepassxc1992.mp4).
Any thoughts on this @varjolintu ?

**Comment by varjolintu** [comment]
@schwedenmut The extension does not yet have any support for detecting overlays. I'm working on it.

**Comment by phoerious** [comment]
That said: all this is pure mitigation for a much larger vulnerability: XSS on the site. An attacker could just create a new login form and let the user fill that in (like "your session has expired, please confirm you login credentials to continue" or so). All the clickjacking stuff is just trying to be sneaky.

**Comment by rugk** [comment]
And thankfully compared to other password managers you don't have a api/gui to show (or heck export) all your passwords *in the website* so a whole range of vulnerabilities gets excluded.
(just from someone coming here from news coverage to clarify that)

If I understand it correctly only the password/credentials for the current size can be stolen in case of an xss from the password manager.

I agree fixing this is good when possible (see below), but indeed if you have a login page on your page and an xss there an attacker could alos just read the input fields of the user anyway wanted to login.
That said click jacking makes this possible on any page of the domain on any element/button without the user knowing its a login at all. This is indeed convenient for an attacker

TL;DR: This is much less serious than other (mostly commercial apparently) password manager that allow(ed) you to export other credentials for other sites too. So for keepassxc this only affects the current site and also only if that has an xss vulnerability in the first place.

**Comment by varjolintu** [comment]
Just a reminder that all these methods require that user has disabled the normal access confirm dialog, and remembered the allowance setting for that single credential for that specific site. By using default settings none of the clickjacking methods will work.

**Comment by marektoth** [comment]
Ok, I will have to respond to the comments because there is a lot of misleading information here.

### **1) You are responding to a clickjacking issue that was opened on July 12, 2021** - unrelated to my latest security research.

So you have had a clickjacking vulnerability in KeePassXC-Browser for over 4 years. The issue was marked as "open" and was discoverable the entire time (I acknowledge my mistake that I should have reported it differently). **For more than 4 years, users were actually exposed to risk** - even the exploit code from test pages is discoverable. In my research, there are companies that didn't fix bugs for 4 months, but you didn't fix a very basic bug even after 4 years - not the various variants I describe in the research.

The fix was made 4 days ago (August 19). Why only now? Didn’t the vulnerability bother you for the past 4 years?

According to the status (https://x.com/KeePassXC/status/1958741619785572495), it might appear that you thought this clickjacking vulnerability was from my latest security research. If you read the entire research, it states: "Extension Element: This is the most basic type, discovered in November 2023, which initiated this entire research. I initially reported only this variant. Later, in early 2025, I identified two additional types and documented the complete DOM-based technique."

TL;DR:
- **You only care about the clickjacking vulnerability after media attention, it didn't bother you for 4 years until now.**
- This ticket doesn't address my latest security research - any fix from the description won't address DOM-based extension clickjacking and their variants.


### 2) @phoerious : "all this is pure mitigation for a much larger vulnerability: XSS on the site. An attacker could just create a new login form and let the user fill that in (like "your session has expired, please confirm you login credentials to continue" or so). All the clickjacking stuff is just trying to be sneaky."

Yes, an attacker could create a new login form, but that is a COMPLETELY DIFFERENT ATTACK VECTOR.
This is a very common mistake and it's necessary to separate the method by which an attacker obtains data and what the success rate of such an attack is.

**Comparison of number of interactions - number of interactions by the victim on the vulnerable website:**
**DOM-based extension clickjacking (KeePassXC-Browser):**
1 click (anywhere) - attacker has login credentials
2 clicks (anywhere) - attacker has login credentials + TOTP

**Copy/paste credentials and TOTP:**
1. Right mouse click (CTRL)
2. Select "Copy username"  (B)
3. Right mouse click (CTRL)
4. Select Paste (V) Username 
5. Right mouse click (CTRL)
6. Select "Copy password" (C)
7. Right mouse click (CTRL)
8. Select Paste (V) Password 	# 8 interactions vs 1 click
9. Right mouse click (CTRL)
10. Select "Copy TOTP" (T)
11. Right mouse click (CTRL)
12. Select Paste TOTP (V)	# 12 interactions vs 2 clicks

- and I'm not counting switching between windows

What if the victim changes their mind during the process? Then the attacker will not get access to the victim account.
The user decides what to enter where and the form is completely visible. **It depends on the user** = it's a social engineering attack

One click anywhere on the page and the user’s credentials are stolen (the user does not see the form or the UI.) = not user mistake, it's a vulnerability in the password manager (=exploiting vulnerability, "social engineering" is minimal - only a single click)

**Result:**
When exploiting clickjacking vulnerability, the attacker needs significantly fewer interactions from the victim

**Attack success rate**

**1.**
When will the attacker have greater attack success rate?
When the user only has to click once anywhere on the page?
Or when the user has to enter all their credentials into a new form?

Definitely, just one click will have greater success rate.

**2.**
If you see a login form anywhere on a page, do you ALWAYS enter your credentials into it? What if you're already logged in, would you still enter your credentials again?
What about clicking on an annoying cookie banner or login dialog, which prevents you from seeing any content on the page? Won't this happen more often? If you want to see the content, you have no choice but to click.

How many times have you logged into GitHub today? And how many times have you clicked on something on GitHub? I bet you close login dialogs or cookies more times during the day than you enter login credentials. 

An attacker can use an exploit containing all vulnerable password managers and browsers - this also significantly increases the chance of success. If I hadn't reported the vulnerabilities, we would now be talking about 12 password managers (including KeePassXC-Browser) and 6 browsers (https://marektoth.com/blog/password-managers-autofill/). I'm also counting automatic autofill (0-click autofill). An attacker could have a script containing exploits for 18 applications and obtain stored credentials with one mouse click from the user (2 clicks only LastPass).

**Result:**
Just one click on a cookies banner will have significantly greater attack success rate than displaying a new form, because with a new form the user is not forced to fill in the credentials (or is forced to, but doesn't have to do it).

An attacker can first use a clickjacking exploit and if they don't gain anything from the click, then they can start their social engineering attack. 

**Can an attacker steal a session using XSS?**
No, it's not that simple. In most cases, the website will have HttpOnly set for session cookies = it's not possible to steal the session using javascript. And even if it were possible, **session != credentials**. Using credentials, an attacker can create a completely new own session - it won't be revoked when the victim logs out.

**TL;DR:**
* If XSS exists and I click just once, nothing happens. But if I'm using a vulnerable password manager and I click = my credentials are stolen
*  **My data is at greater risk if I use a vulnerable password manager than if I don’t use one at all** (the webadmin has no responsibility that I use a vulnerable password manager) ⚠️

* **Can an attacker obtain a user’s login credentials via XSS with just one click from the victim? No, the attacker cannot get them.** ⚠️
* I’m not concerned with everything that can be done via XSS in the webapp - my goal is plaintext credentials with a greater success rate

### 3) @rugk : This is much less serious than other (mostly commercial apparently) password manager that allow(ed) you to export other credentials for other sites too.

Export credentials was described in the IFRAME-based variant (web_accessible_resources), but that wasn't my main research. I even stated this: "I will first describe the IFRAME-based variant, which was not the research focus but may be unknown to many people." KeePassXC-Browser is vulnerable to DOM-based extension clickjacking.

### 4) @varjolintu : Just a reminder that all these methods require that user has disabled the normal access confirm dialog, and remembered the allowance setting for that single credential for that specific site. By using default settings none of the clickjacking methods will work.

That's not true ⚠️ 

Please do not write false information, as you are misleading people. People expect you to be the developer of the application, so they will assume the statements are verified.

If user data is stored via browser, then the access confirm dialog doesn't show up ⚠️ 
And if the user has KeePassXC-Browser (browser integration), why would they always store data via KeePassXC?

Video with clean installation with default settings:
https://youtu.be/Mx0YWLH6pYA

Btw. your comment from Jul 13, 2021 on this issue:

> "Ah yes. If you create an entry using the extension, the host is automatically allowed. Maybe we should make that optional. If you just create an entry using KeePassXC, the confirmation dialog will be shown."

----

**Is KeePassXC-Browser vulnerable to DOM-based Extension Clickjacking:** YES
**Is it vulnerable in default settings:** YES (data stored via KeePassXC-Browser)
**Vulnerable version:** <=1.9.9.2 (latest)
**Is XSS vulnerability needed:** YES - KeePassXC-Browser only fills credentials (as far as I know)

**Comment by phoerious** [comment]
I'm pretty sure the success rate of a visible form would be similar and there isn't anything we can do against that.

This issue was NOT open for four years and we just didn't bother. We fixed several ways someone could click jack users' information via invisible fields (#851, #906, #1368, #2641, #2646). #1368 was in direct response to this issue, the latter two to the new reports. You came up with new ways that we didn't protect against and I'm sure there will be yet other ways you could exploit invisible forms.

**Comment by marektoth** [comment]
This issue was NOT open? Really? 

May 22 2023
Open: https://web.archive.org/web/20230522120836/https://github.com/keepassxreboot/keepassxc-browser/issues/1367

October 02 2024
Open: https://web.archive.org/web/20241002014123/https://github.com/keepassxreboot/keepassxc-browser/issues/1367

> I'm sure there will be yet other ways you could exploit invisible forms.

I am not saying anywhere that the number of ways is final - "If other methods exist or there is some bypass, then the extensions are still vulnerable (but I don't know of any at the moment)."

**Comment by varjolintu** [comment]
> ### 4) [@varjolintu](https://github.com/varjolintu) : Just a reminder that all these methods require that user has disabled the normal access confirm dialog, and remembered the allowance setting for that single credential for that specific site. By using default settings none of the clickjacking methods will work.
> 
> That's not true ⚠️
> 
> Please do not write false information, as you are misleading people. People expect you to be the developer of the application, so they will assume the statements are verified.
> 
> If user data is stored via browser, then the access confirm dialog doesn't show up ⚠️ And if the user has KeePassXC-Browser (browser integration), why would they always store data via KeePassXC?

What I said is true. Please do not write false information, as it might mislead people.

By default KeePassXC prompts the user **every time** a credential is matched before returning it to the extension. If user has remembered the allowance setting in the Access Confirm dialog, only after that extension data is stored to the entry. Without checking the "Remember" option, user is not vulnerable to any of the clickjacking methods unless the user manually allows returning the entry from KeePassXC (and this is required with every access).

The dialog:
<img width="522" height="365" alt="Image" src="https://github.com/user-attachments/assets/a6c9f4b4-832b-414e-8ed9-d62fbc228c87" />

**Comment by marektoth** [comment]
Screenshot is proof that you stored credentials via KeePassXC-Browser? Please record video with your steps. 

Did you see my video? Could you explain **why it was autofilled for me without requiring confirmation**? The whole process from installation to autofill is recorded.

**Comment by phoerious** [comment]
> This issue was NOT open? Really?
> 
> May 22 2023 Open: https://web.archive.org/web/20230522120836/https://github.com/keepassxreboot/keepassxc-browser/issues/1367
> 
> October 02 2024 Open: https://web.archive.org/web/20241002014123/https://github.com/keepassxreboot/keepassxc-browser/issues/1367

The original issue was fixed back then and the issue was closed. It was reopened at some point for the additional overlay method. You can see that in the history of this issue, you don't need the Internet Archive to see that. Should we have fixed that earlier? Yes, probably. But since KeePassXC asks for each entry by default (unless you saved it via browser extension, which is something we should fix), you cannot retrieve any credentials that aren't for that page, and a proper fix isn't trivial, it was rather low priority.

**Comment by varjolintu** [comment]
> Screenshot is proof that you stored credentials via KeePassXC-Browser? Please record video with your steps.
> 
> Did you see my video? Could you explain **why it was autofilled for me without requiring confirmation**? The whole process from installation to autofill is recorded.

This is because currently we allow the site automatically when a new credential is saved using the extension. However, change to this has been requested earlier and we're gonna change that. Entries created manually to KeePassXC are not affected.

**Comment by marektoth** [comment]
I'm ending this discussion. I really can't with this. Now, you're literally confirming what I wrote.

![](https://marektoth.com/images/comment.png)

**Comment by varjolintu** [comment]
@marektoth I meant the default setting is that you have credentials already in KeePassXC, not stored via the extension.

**Comment by marektoth** [comment]
> Just a reminder that all these methods require that user has disabled the normal access confirm dialog, and remembered the allowance setting for that single credential for that specific site. By using default settings none of the clickjacking methods will work.

Where is KeePassXC mentioned? Vulnerablity is in KeePassXC-Browser as I wrote:

> If user data is stored via browser, then the access confirm dialog doesn't show up ⚠️
>Is KeePassXC-Browser vulnerable to DOM-based Extension Clickjacking: YES
>Is it vulnerable in default settings: YES (data stored via KeePassXC-Browser)

What does this mean "By using default settings none of the clickjacking methods will work."? 

The entire my comment was about default settings in browser extension - KeePassXC-Browser. My research is about the browser extension. This is KeePassXC-Browser repository. And you start writing that KeePassXC (without mention) is not vulnerable in default settings? And then you say "Please do not write false information, as it might mislead people", when I’m only talking about KeePassXC-Browser? Really?

**Comment by varjolintu** [comment]
@marektoth Just pointing the difference with entries in created in KeePassXC vs. created using the extension. Many people might think both are affected.

**Comment by phoerious** [comment]
I don't think we need to discuss semantics here. You identified a hole in our mitigation heuristics (thanks!) and we're fixing it from multiple angles.

**Comment by PhrozenByte** [comment]
https://github.com/keepassxreboot/keepassxc-browser/issues/1367#issuecomment-3217919415

> By default KeePassXC prompts the user every time a credential is matched before returning it to the extension. If user has remembered the allowance setting in the Access Confirm dialog, only after that extension data is stored to the entry. Without checking the "Remember" option, user is not vulnerable to any of the clickjacking methods unless the user manually allows returning the entry from KeePassXC (and this is required with every access).

@varjolintu, I'm sorry, but from an user's and fellow developer's perspective this kinda feels like a weak excuse. Nowhere in KeePassXC nor KeePassXC-Browser is an indication that using the "Remember" checkbox could pose a security threat. From an UX perspective it doesn't make much sense either: This dialogue asks whether a particular entry in the password database is meant to be used by a particular browser (or: browser addon) for a particular website. There's no indication that this is supposed to be used on a per-access basis - because, if it would, why does the "Remember" checkbox even exist?

---

In general, clickjacking is a well-established attack vector that needs to be addressed. Some users excuse this by applying Whataboutism ("What about showing the user a login form instead"), but this simply is no valid argument. In security there's a huge difference between social engineering an user to do something, or just doing it without or with unrelated user interaction.

Still, this security issue is no CVSS 10.0. It's not the most critical issue ever seen in software, but downplaying it like "we're not affected in default config" is neither a good excuse in general, but especially not if it isn't even true in its full extent, but only semantically. One wouldn't argue that KeePassXC isn't storing any credentials in its default config and thus no credentials can ever be stolen in its default config either…

This is an open source project, so nobody is expecting anyone to address every finding right away, especially not sacrificing one's free time at a weekend. But downplaying it is bad on its own. Please, just don't do that. It leaves a bitter aftertaste, even after the issue has been resolved. From my understanding @varjolintu is already working it. That's amazing news! Thank you @varjolintu! :heart:

It's a security researcher's job to find attack vectors nobody thought about before. Security researchers aren't attacking developers by reporting issues, they rather aim to help to improve things and do good. It's a very good sign that security researches actually care about KeePassXC and do independent research on it. We should all be happy that things are improving! Security researchers are no offenders. Please don't make security researcher feel like offenders.

@marektoth Thank you for your security research, not just on KeePassXC, but the many password managers out there! :heart:

**Comment by phoerious** [comment]
FYI https://github.com/keepassxreboot/keepassxc/pull/12413

**Comment by Natha** [comment]
I'll make it "short". 

1. [Consent Fatigue](https://iapp.org/news/a/how-to-avoid-consent-fatigue) is a real thing. There's a deeper analysis about it available [here](https://www.edpb.europa.eu/sites/default/files/webform/public_consultation_reply/fedoweb-_-consultazione-pubblica-edpb-art-5.3-eprivacy-directive_0.pdf). If it relies on a technical vulnerability that can be fixed, it cannot be considered (and so dismissed, which is what 1Password did) as social engineering. 

2. Everything that @marektoth has said is true and can be verified by anyone through the demo website. The [screen recording](https://youtu.be/Mx0YWLH6pYA?si=2MHq_lyQaz36mbnL) he shared is as straightforward as it can be.

3. Everyone makes mistake. Bugs and leaks are most of the time due to human weaknesses of any kind. A good old [rm -rf](https://youtu.be/tLdRBsuvVKc?si=amEUPSmGuwsKoYFS) in the wrong location is a pretty good example of that.

I'm not here to replicate the Spider-Man pointing at Spider-Man meme. Neither am I looking to cause even more trouble in a situation that is already complicated for every password manager developer. I sincerely think tho that many in the dev community, both the vendors (and their dev team) and the indie devs should stop minimising the vulnerabilities discovered by @marektoth .

To every brand/companies/software their devotees. It's not gonna change anytime soon. But, please, for the love of Vim, don't forget that not every user is a professional or a tech-savvy person. Not everyone will dig through the settings of your apps. It's already a miracle that a common person is using _any_ kind of safe password manager. And even that being said, these vulnerabilities could also impact an experienced user.

If there isn't a notification sent to the user that this and that happened (in this case, your password/credit card/ID/etc.), it is something that needs to be fixed. If an innocent cookie overlay ends up being an antipersonnel landmine,then problems are bound to happen in a disastrous way.  

These security issues have to be taken seriously. So that more and more people decide to use a password manager to protect themselves from the exponentially growing attacks & leaks of so many websites/services/whatever. And so that the current users of these password managers don't feel betrayed or (worse) patronised by a company they are paying for to be safe in the first place.

I am not pointing the finger at KeePass and its team. You're maintaining an open source software. Meaning, we can only assume that you're doing your best. 

My anger is very much expressed towards the big ones like 1Password, which didn't deal with this the right way.

It doesn't change the fact that every one should take a step back sit down, and realise that there's an issue to fix. Not in 4 months. Not in under a year. Or worse. 

Long story short: yes, the users are responsible of their own safety. Most of the time, shit happens because of a problem that is between the chair and the desk. But devs working on security-oriented software are still very much responsible of their users safety. The "not in my yard" mentality of some devs over the last few days has been unacceptable. And I do hope that this will change for the better in the future. 

Good luck to y'all. Thanks for your work on these things. Let's all learn something out of them and move on.

**Comment by phoerious** [comment]
I'm locking this, since everything has been said. We're only asking that you keep things in proportion. There's a difference between an arbitrary page being able to extract *all* passwords and an already hijacked page being able to extract its own credentials. These are entirely different things and should be treated that way. Unlike some of the other password managers, KeePassXC isn't vulnerable to the former and never was. We also don't fill credit card info or other global data. Please keep separate things separate. This here is not a CVSS 10, nor would fixing this particular thing prevent XSS-pwned pages from stealing its own credentials (which doesn't mean it shouldn't be fixed and we are working on it).
